src/main/java/org/apache/sysds/runtime/matrix/data/LibMatrixFourier.java

```diff
@@ -440,4 +444,18 @@ public static MatrixBlock[] ifft_linearized(MatrixBlock re, int threads) {
 			threads);
 	}
 
+	private static double sin(double angle){
+		if (!sinCache.containsKey(angle)) {
```

Comment:
please limit the cache to contain only a subset of values (max 1000), since the map can grow to large.
also we need to hide it behind a SoftReference since we want to allow java to remove it from memory if we are under memory pressure.

Therefore to use it extract the HashMap and materialize it as a variable inside fft_one_dim_sub(), and put that hashmap in as an argument to your calls to cos and sin.

Reply:
Would it also be fine to materialize the cache inside of fft() or fft_one_dim()? As far as I can see, cos or sin is never computed for the same values in fft_one_dim_sub() and therefore there would be no benefit.

Reply:
I think my comment was unclear.

1. You should reuse the same cached hash map every invocation.
2. The soft reference is just wrapping your allocated caches.
3. Because it takes extra time to get your cached angles (the HashMap) out of the soft reference, we need to get it out at a different location. Therefore i suggested to materialize it as a variable inside "fft_one_dim_sub", meaning taking it out of the soft reference there, or materializing a new hash map if the soft reference is empty.